### ISSUE TITLE
Chinese Text Selection Issue in SumatraPDF with Files Created Using Typst (Specific Fonts)

### ISSUE THREAD

**Issue by brynne8** [sumatrapdfreader/sumatrapdf#3497]
I am using the latest version of SumatraPDF and have encountered an issue with selecting Chinese text in PDF files created with [Typst](https://github.com/typst/typst). The text selection behavior for Chinese characters is not functioning as expected when using certain fonts. Specifically, the following problems are observed:

1. It is not possible to initiate text selection by clicking on a Chinese character when using fonts other than SimSun.
2. When starting text selection from a Latin character, the Chinese text does not appear as selected if it is not in the SimSun font.

[test.pdf](https://github.com/sumatrapdfreader/sumatrapdf/files/11485040/test.pdf)

![img](https://github.com/sumatrapdfreader/sumatrapdf/assets/7542439/579906b8-bdfc-4a85-a068-57b5e68156fb)

It is worth noting that Chinese text using the SimSun font works as expected, and the issue is not present. However, for other fonts, these issues make it difficult to work with documents containing Chinese text in SumatraPDF. I have not encountered this problem with other PDF readers. Please investigate this issue and provide a fix or workaround to improve the text selection experience for Chinese characters in SumatraPDF when using various fonts.

Thank you for your attention to this matter.

**Comment by GitHubRulesOK** [comment]
Yes this is a problem with many fonts, not just CJK but plain English or European that they need to be a glyph of uniform em sized metrics so ideally  like this.

```
自动化构建最快的Lua解释器!
Font KaiTi example: 动态语言的虚拟机实现并非易事，众所周知。像Safari中的JavaScript解释
器、LuaJIT中的Lua解释器等高性能的解释器，通常都是手工编写的汇编代码。如果你想要通
过JIT编译来提高性能，那你需要编写更多的汇编代码。如果你想要使用多级JIT编译来达到最
好的性能……好吧，那就是从头到尾都是汇编代码。
Font SimSun example: 我一直在做一个实现虚拟机更简单的研究项目。想法来源于下面这种观
察:编写一个简单的解释器并不困难（就是写一个大的switch-case），但是编写一个优秀的解释
器（或JIT编译器）却很困难，因为它不可避免地要涉及手写汇编代码。那么为什么我们不能
实现一个特殊的编译器，可以从“这个大switch-case”，或者更正式地说，从每个字节码的语义
描述中，自动生成出高性能的解释器（甚至JIT解释器）
```

**Comment by YDX-2147483647** [comment]
Sorry for disturbing… I've just fixed another issue https://github.com/typst/typst/issues/3416 on PDF generated by Typst with specific fonts. Hope my experience helps.

It's likely that Typst gives wrong/insufficient information for SumatraPDF. If anyone have time to investigate, you can upload a minimal working example to [PDF Object Browser](https://brendandahl.github.io/pdf.js.utils/browser/). It is one of the debugging tools mentioned in [a post on Typst Blog](https://typst.app/blog/2023/color-gradients#debugging-pdf). Besides, you can also refer to [PDF Cheat Sheets – PDF Association](https://pdfa.org/resource/pdf-cheat-sheets/) and [ISO 32000 (PDF) – PDF Association](https://pdfa.org/resource/iso-32000-pdf/) for detailed PDF spec with examples. §9 of [PDF32000_2008.pdf](https://opensource.adobe.com/dc-acrobat-sdk-docs/pdfstandards/PDF32000_2008.pdf) is the most pertinent.

Anyway, live long and prosper from the new Chinese Loong🐉 year! 龙年大吉！

**Comment by Vanishine** [comment]
@GitHubRulesOK @YDX-2147483647 

When I use [ConTeXt](https://wiki.contextgarden.net/Main_Page), I encountered the issue you mentioned. I generated two PDFs using the same font, "[IBM Plex Sans SC](https://github.com/IBM/plex/releases/tag/%40ibm%2Fplex-sans-sc%401.1.0)", in two different formats: TTF and OTF. The PDF generated with the OTF version had the problem, while the TTF version did not. Interestingly, when I switched to the "KaiTi" font, the issue wouldn't show up.

This contrasts with Typst 0.11.0, where the same problem occurred with "KaiTi" but was resolved in Typst 0.12.0 (though I am unsure which specific change fixed it).

I also tested "[Source Han Sans](https://github.com/adobe-fonts/source-han-sans/tree/release?tab=readme-ov-file#otcs)" (static OTCs), which exhibited the same issue in ConTeXt.

| IBM Plex Sans SC (OTF) | IBM Plex Sans SC (TTF) | KaiTi (TTF) |
|--------|--------|--------|
| [ibm-otf.pdf](https://github.com/user-attachments/files/18052011/ibm-otf.pdf) | [ibm-ttf.pdf](https://github.com/user-attachments/files/18052013/ibm-ttf.pdf) | [kaiti.pdf](https://github.com/user-attachments/files/18052014/kaiti.pdf) |
| ![IBM Plex Sans SC (OTF)](https://github.com/user-attachments/assets/76313fff-e65a-4c91-8352-a1443ed023d9) | ![IBM Plex Sans SC (TTF)](https://github.com/user-attachments/assets/4912ca36-4bb0-41d2-ae74-b307ed2a7057) | ![KaiTi](https://github.com/user-attachments/assets/3b7799f7-bdde-4e45-8c2e-3ce48bc5dcf0) | 

```tex
\starttext
\startTEXpage[offset=12pt]
\definefont[han][name:ibmplexsanssc*default at 12pt]
%\definefont[han][name:kaiti*default at 12pt]
\han
喜上眉梢
\stopTEXpage
\stoptext
```

**Comment by YDX-2147483647** [comment]
> the same problem occurred with "KaiTi" but was resolved in Typst 0.12.0

I can verify it!

The following items in [typst v0.12.0 changelog](https://typst.app/docs/changelog/0.12.0/#export) might be relevant.
> - Highly reduced PDF file sizes due to better font subsetting
> - Fixed a bug where some fonts would not print correctly on professional printers

Considering it, I suggest we change the issue title?

**Comment by GitHubRulesOK** [comment]
Closing as OP file is now seen to be as expected in latest updated pre-release

![Image](https://github.com/user-attachments/assets/633ea87a-5e25-42b8-8f54-356e0eab03c4)
